### PR TITLE
alias support for ORCA

### PR DESCRIPTION
**Pull request by dkovalev1**
Add support for table aliases in ORCA

Currently ORCA generated query plan does not show aliases, unlike original
PostgreSQL planner. Plan with no aliases can look different from the original
query, what makes it difficult to read for some users.

This change adds table aliases to ORCA. Table aliases are added to both internal
structures and DXL schema.

This change amends a lot of plans generated in existing test suite, so it seems
not feasible to add specialized test for it.

### CODE REVIEW

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/62930

**Comment by BenderArenadata** [comment]
Failed job ORCA code linter on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1016262

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1016260

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1016261

File: src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp [hunk 1]
```diff
@@ -7553,10 +7553,15 @@ CTranslatorExprToDXL::MakeDXLTableDescr(const CTableDescriptor *ptabdesc,
 
 	CMDIdGPDB *mdid = CMDIdGPDB::CastMdid(ptabdesc->MDId());
 	mdid->AddRef();
-
+
```
Comment (RekGRpth) [inline src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp]: hmm...

File: src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp [hunk 2]
```diff
@@ -7553,10 +7553,15 @@ CTranslatorExprToDXL::MakeDXLTableDescr(const CTableDescriptor *ptabdesc,
 
 	CMDIdGPDB *mdid = CMDIdGPDB::CastMdid(ptabdesc->MDId());
 	mdid->AddRef();
-
+                                                                                                                                                                                                                       
 	CDXLTableDescr *table_descr = GPOS_NEW(m_mp)
 		CDXLTableDescr(m_mp, mdid, pmdnameTbl, ptabdesc->GetExecuteAsUserId());
 
+	if (ptabdesc->IsAliasUsed()) {
```
Comment (RekGRpth) [inline src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp]:
```suggestion
	if (NULL != ptabdesc->Alias()) {
```

Comment (dkovalev1) [reply src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp]: This won't work, CTableDescriptor::Alias() returns reference, not a pointer. Basically, this is a reason for introducing CTableDescriptor::IsAliasUsed().

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1016254

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1016255

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/63018

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1020493

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1020494

**Comment by BenderArenadata** [comment]
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1020492

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1020485

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1020487

File: src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp [hunk 3]
```diff
@@ -4672,8 +4672,11 @@ CTranslatorDXLToPlStmt::TranslateDXLTblDescrToRangeTblEntry(
 	alias->colnames = NIL;
 
 	// get table alias
+	const CMDName *md_alias =
+		table_descr->MdAlias() ? table_descr->MdAlias() : table_descr->MdName();
```
Comment (RekGRpth) [inline src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp]:
```suggestion
		(NULL != table_descr->MdAlias() ? table_descr->MdAlias() : table_descr->MdName());
```

File: src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp [hunk 4]
```diff
@@ -4724,6 +4727,12 @@ CTranslatorDXLToPlStmt::TranslateDXLTblDescrToRangeTblEntry(
 
 	rte->eref = alias;
 
+	if (NULL != md_alias)
+	{
+		rte->alias = alias;
```
Comment (RekGRpth) [inline src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:4737]: What does this condition mean?

File: src/backend/gpopt/translate/CTranslatorUtils.cpp [hunk 5]
```diff
@@ -136,6 +136,16 @@ CTranslatorUtils::GetTableDescr(CMemoryPool *mp, CMDAccessor *md_accessor,
 	CDXLTableDescr *table_descr =
 		GPOS_NEW(mp) CDXLTableDescr(mp, mdid, table_mdname, rte->checkAsUser);
 
+	if (rte->alias != NULL)
```
Comment (RekGRpth) [inline src/backend/gpopt/translate/CTranslatorUtils.cpp]:
```suggestion
	if (NULL != rte->alias)
```

File: src/backend/gporca/libnaucrates/src/operators/CDXLTableDescr.cpp [hunk 6]
```diff
@@ -86,6 +87,26 @@ CDXLTableDescr::MdName() const
 	return m_mdname;
 }
 
+
+// table alias accessors
+const CMDName *
+CDXLTableDescr::MdAlias() const
```
Comment (RekGRpth) [inline src/backend/gporca/libnaucrates/src/operators/CDXLTableDescr.cpp:93]: why not put them into h-file like CTableDescriptor.h?

File: src/backend/gporca/libnaucrates/src/operators/CDXLTableDescr.cpp [hunk 7]
```diff
@@ -86,6 +87,26 @@ CDXLTableDescr::MdName() const
...
+{
+	return m_mdalias;
+}
+
+void
+CDXLTableDescr::SetMdAlias(CMDName *alias)
+{
+	if (m_mdalias != NULL)
```
Comment (RekGRpth) [inline src/backend/gporca/libnaucrates/src/operators/CDXLTableDescr.cpp]:
```suggestion
	if (NULL != m_mdalias)
```

File: src/backend/gporca/libnaucrates/src/operators/CDXLTableDescr.cpp (hunk 6 above)
Comment (dkovalev1) [reply src/backend/gporca/libnaucrates/src/operators/CDXLTableDescr.cpp:93]: To keep consistent with CDXLTableDescr::MdName() which is also has a separated declaration. Personally I'd prefer to have in inline in header too.

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/63087

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1024402

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1024403

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1024396

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1024397

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/63121

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1026560

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1026561

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1026555

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1026554

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/63256

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1032790

**Comment by BenderArenadata** [comment]
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1032788

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1032791

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1032784

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1032785

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/63459

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1042571

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1042572

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1042562

File: src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp [hunk 8]
```diff
@@ -4724,6 +4728,16 @@ CTranslatorDXLToPlStmt::TranslateDXLTblDescrToRangeTblEntry(
 
 	rte->eref = alias;
 
+	/*
+		Store the alias information only when table alias is used to allow 
+		EXPLAIN find it
+	*/
+	if (NULL != md_alias)
```
Comment (RekGRpth) [inline src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:4735]: [line](https://github.com/arenadata/gpdb/pull/682/files#diff-2dcfa821b8990b5ae39efff4a97497121dd2e40c31f67d6f180d228f1f34b2c8R4680) dereferences md_alias

File: src/backend/gporca/libgpopt/include/gpopt/metadata/CTableDescriptor.h [hunk 9]
```diff
@@ -54,6 +54,9 @@ class CTableDescriptor : public CRefCount,
 	// name of table
 	CName m_name;
 
+	// table alias
+	CName *m_alias;
```
Comment (RekGRpth) [inline src/backend/gporca/libgpopt/include/gpopt/metadata/CTableDescriptor.h:58]: How about changing the type of this field to CWStringConst? Then below, in the SetAlias method there will be no need for a constructor, and in the destructor there will be no need for deletion and nulling.

File: src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp (hunk 8 above)
Comment (dkovalev1) [reply src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:4735]: It does. Initialization of md_alias makes sure that it is not NULL.

File: src/backend/gporca/libgpopt/include/gpopt/metadata/CTableDescriptor.h (hunk 9 above)
Comment (dkovalev1) [reply src/backend/gporca/libgpopt/include/gpopt/metadata/CTableDescriptor.h:58]: m_alias is CName to be consistent with declaration of m_name above. Pointer is needed to make it optional. Changing it to CWStringConst pointer will result in using similar constructor and destructor and using an instance will require additional boolean optional flag and more complicated logic.

**Comment by RekGRpth** [comment]
may be close this?

**Pull request closed without merging**